internal/builtin/providers/tf/provider_functions.go

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+
+package tf
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2/hclwrite"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/opentofu/opentofu/internal/providers"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// "encode_tfvars"
+// "decode_tfvars"
+// "encode_expr"
+
+type providerFunc interface {
+	// Name returns the name of the function which is used to call it
+	Name() string
+	// GetFunctionSpec returns the provider function specification
+	GetFunctionSpec() providers.FunctionSpec
+	// Call is used to invoke the function
+	Call(args []cty.Value) (cty.Value, error)
+}
+
+// getProviderFuncs returns a map of functions that are registered in the provider
+func getProviderFuncs() map[string]providerFunc {
+	decodeTFVars := &decodeTFVarsFunc{}
+	encodeTFVars := &encodeTFVarsFunc{}
+	encodeExpr := &encodeExprFunc{}
+	return map[string]providerFunc{
+		decodeTFVars.Name(): decodeTFVars,
+		encodeTFVars.Name(): encodeTFVars,
+		encodeExpr.Name():   encodeExpr,
+	}
+}
+
+// decodeTFVarsFunc decodes a TFVars file content into a cty object
+type decodeTFVarsFunc struct{}
+
+func (f *decodeTFVarsFunc) Name() string {
+	return "decode_tfvars"
+}
+
+func (f *decodeTFVarsFunc) GetFunctionSpec() providers.FunctionSpec {
+	params := []providers.FunctionParameterSpec{
+		{
+			Name:              "content",
+			Type:              cty.String,
+			Description:       "TFVars file content to decode",
+			DescriptionFormat: providers.TextFormattingPlain,
+		},
+	}
+	return providers.FunctionSpec{
+		Parameters:        params,
+		Return:            cty.DynamicPseudoType,
+		Summary:           "Decode a TFVars file content into an object",
+		Description:       "provider::terraform::decode_tfvars decodes a TFVars file content into an object",
+		DescriptionFormat: providers.TextFormattingPlain,
+	}
+}
+
+var FailedToDecodeError = errors.New("failed to decode tfvars content")
+
+func wrapDiagErrors(m error, diag hcl.Diagnostics) error {
+	//Prepend the main error
+	errs := append([]error{m}, diag.Errs()...)
+	return errors.Join(errs...)
+}
+
+func (f *decodeTFVarsFunc) Call(args []cty.Value) (cty.Value, error) {
+	varsFileContent := args[0].AsString()
+	schema, diag := hclsyntax.ParseConfig([]byte(varsFileContent), "", hcl.Pos{Line: 0, Column: 0})
```

Comment:
Does that function support non vars file formats? We might want to additionally check if that's the case and error for non-supported files

Reply:
Not anymore no, thanks for catching that!

Reply:
Do you refer to the file extension? 
They pass in the file's content, not the actual file. ParseConfig ensures the content is in the correct HCL syntax. 
Did I understand your question correctly?

Reply:
Yep, I thought the `hclsyntax.ParseConfig` could parse something more generic than just a var file syntax. I double checked other usages and it seems fine to me.